src/Agent.Worker/CodeCoverage/CodeCoverageServer.cs

```diff
@@ -29,7 +32,35 @@ public async Task PublishCoverageSummaryAsync(VssConnection connection, string p
                 CoverageStats = coverageData.ToList()
             };
 
-            await testHttpClient.UpdateCodeCoverageSummaryAsync(data, project, buildId, cancellationToken: cancellationToken);
+            FeatureAvailabilityHttpClient featureAvailabilityHttpClient = connection.GetClient<FeatureAvailabilityHttpClient>();
+            if (GetFeatureFlagState(featureAvailabilityHttpClient, CodeCoverageConstants.EnablePublishToTcmServiceDirectlyFromTaskFF, context))
+            {
+                TestResultsHttpClient tcmClient = connection.GetClient<TestResultsHttpClient>();
+                await tcmClient.UpdateCodeCoverageSummaryAsync(data, project, buildId, cancellationToken: cancellationToken);
+            }
+            else
+            {
+                TestManagementHttpClient tfsClient = connection.GetClient<TestManagementHttpClient>();
+                await tfsClient.UpdateCodeCoverageSummaryAsync(data, project, buildId, cancellationToken: cancellationToken);
+            }
+        }
+
+        private bool GetFeatureFlagState(FeatureAvailabilityHttpClient featureAvailabilityHttpClient, string FFName, IAsyncCommandContext context)
+        {
+            try
+            {
+                var featureFlag = featureAvailabilityHttpClient?.GetFeatureFlagByNameAsync(FFName).Result;
+                if (featureFlag != null && featureFlag.EffectiveState.Equals("On", StringComparison.OrdinalIgnoreCase))
+                {
+                    return true;
+                }
+            }
+            catch
+            {
+                context.Debug(StringUtil.Format("Failed to get FF {0} Value. By default, publishing data to TCM.", FFName));
+                return true;
+            }
+            return false;
         }
```

Comment:
Suggest to refactor this guy to its own file. this will ensure single responsibility model of classes are taken care of